libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -421,6 +436,113 @@ struct ReservedErrorSelector: public PostTypeChecker::Checker
+};
+
+class LValueChecker: public ASTConstVisitor
+{
+public:
+	LValueChecker(Identifier const& _identifier):
+		m_declaration(_identifier.annotation().referencedDeclaration)
+	{}
+	bool willBeWrittenTo() const { return m_willBeWrittenTo; }
+	void endVisit(Identifier const& _identifier) override
+	{
+		solAssert(_identifier.annotation().referencedDeclaration);
+		if (
+			*_identifier.annotation().referencedDeclaration == *m_declaration &&
+			_identifier.annotation().willBeWrittenTo
+		)
+			m_willBeWrittenTo = true;
+	}
+	void endVisit(InlineAssembly const& _inlineAssembly) override
+	{
+		YulLValueChecker yulChecker{m_declaration->name()};
+		yulChecker(_inlineAssembly.operations());
+		m_willBeWrittenTo = yulChecker.willBeWrittenTo();
+	}
+private:
+	Declaration const* m_declaration{};
+	bool m_willBeWrittenTo = false;
+};
+
+struct SimpleCounterForLoopChecker: public PostTypeChecker::Checker
+{
+	SimpleCounterForLoopChecker(ErrorReporter& _errorReporter): Checker(_errorReporter) {}
+	bool visit(ForStatement const& _forStatement) override
+	{
+		_forStatement.annotation().isSimpleCounterLoop = isSimpleCounterLoop(_forStatement);
+		return true;
+	}
+	bool isSimpleCounterLoop(ForStatement const& _forStatement) const
+	{
+		auto const* simpleCondition = dynamic_cast<BinaryOperation const*>(_forStatement.condition());
+		if (!simpleCondition || simpleCondition->getOperator() != Token::LessThan || simpleCondition->userDefinedFunctionType())
+			return false;
+		if (!_forStatement.loopExpression())
+			return false;
+
+		auto const* simplePostExpression = dynamic_cast<UnaryOperation const*>(&_forStatement.loopExpression()->expression());
+		// This matches both operators ++i and i++
+		if (!simplePostExpression || simplePostExpression->getOperator() != Token::Inc || simplePostExpression->userDefinedFunctionType())
+			return false;
+
+		auto const* lhsIdentifier = dynamic_cast<Identifier const*>(&simpleCondition->leftExpression());
+		auto const* lhsIntegerType = dynamic_cast<IntegerType const*>(simpleCondition->leftExpression().annotation().type);
+		auto const* commonIntegerType = dynamic_cast<IntegerType const*>(simpleCondition->annotation().commonType);
+
+		if (!lhsIdentifier || !lhsIntegerType || !commonIntegerType || *lhsIntegerType != *commonIntegerType)
+			return false;
+
+		if (
+			!dynamic_cast<Identifier const*>(&simpleCondition->rightExpression()) &&
+			!dynamic_cast<Literal const*>(&simpleCondition->rightExpression())
+		)
+		{
+			solAssert(lhsIdentifier);
+			LValueChecker rhsConditionChecker{*lhsIdentifier};
+			simpleCondition->rightExpression().accept(rhsConditionChecker);
+			if (rhsConditionChecker.willBeWrittenTo())
+				return false;
+		}
+
+		auto const* incExpressionIdentifier = dynamic_cast<Identifier const*>(&simplePostExpression->subExpression());
+		if (
+			incExpressionIdentifier &&
+			incExpressionIdentifier->annotation().referencedDeclaration != lhsIdentifier->annotation().referencedDeclaration
```

Comment:
This check is skipped when the increment expression is an identifier. What if it's an expression?

```solidity
uint j = type(uint).max;
for (uint i = 0; i < 10; ++(j)) { ... }
```

Or member access?

```solidity
struct S {
    uint x;
}

...

S memory s = S(type(uint).max);
for (uint i = 0; i < 10; ++s.x) { ... }
```

Not sure if these will really bypass the check - both will probably fail on the next check since `(j)` has no referenced declaration and `s.x` might not count as a local variable. But there might be other situations I haven't thought about.

What you probably need here is to fail the check when the increment expression is not an identifier:

```suggestion
			!incExpressionIdentifier ||
			incExpressionIdentifier->annotation().referencedDeclaration != lhsIdentifier->annotation().referencedDeclaration
```

Reply:
Both of those cases, expressions and member access, were not covered in the tests. Hence, I hadn't gotten in the problem you point here, which is actually more severe, because, as you said they have no referenced declaration, leading to a seg fault.
Fixed.